### DISCUSSION TITLE
Special Character Problem

### DISCUSSION THREAD

**Discussion by Knutsford** [dompdf/dompdf#2891] [Q&A]
I have `&comma;` appearing in the pdf instead of a , even thought I have used htmlspecialchars_decode() to convert it. How do I get a , appearing instead of `&comma;` and the same with any other Special Characters. Thanks

**Comment by bsweeney** [comment]
Which version are you using? Taking a quick look at how that entity renders in various releases I do see the issue starting with release 0.8.4 that runs through 1.2.1. It renders correctly in 1.2.2 (the current release). I'd have to look into it the issue more to see exactly where things broke.

**Comment by Knutsford** [comment]
1.0.2 is the version. Hmm – I thought I had downloaded a newer one than that as it wasn’t that long ago. Ok I will download the latest one in the morning and try it - thanks

**Comment by Knutsford** [comment]
Any luck yet with finding out why the &comma; problem is still happening in 1.2.2?

**Reply by bsweeney** [reply]
Is it still happening on 1.2.2? I did a quick test on the [debug helper](https://eclecticgeek.com/dompdf/debug.php) and it renders OK.

**Comment by Knutsford** [comment]
Yep it is ☹

I have attached what I am getting.

**Comment by Knutsford** [comment]
`<h3">Answer: ' . trim($Answers[$i]) . '</h3>` doesn’t render a comma as part of trim($Answers[$i])

`<h3">,</h3>` renders a comma alright 

`. trim($Answers[$i]) .` is the problem whether it has a `&comma;` or just a , as part of the answer

If that helps

**Reply by bsweeney** [reply]
Would need to see the actual HTML after the PHP is parsed to have a better sense of what's going on.

**Comment by Knutsford** [comment]
Is the HTML:  https://github.com/dompdf/dompdf/files/8782367/sample.txt

Hmm – that is coming correctly when you view it in a browser but not in the PDF.

**Reply by bsweeney** [reply]
It looks like the stray `"` in your H3 element is the cause. This is probably giving the parser a bit of a fit. Take that out and everything should render as expected.

**Comment by Knutsford** [comment]
`&pound;` is ok but not `&comma;`

Eg this is ok

```
html_entity_decode(trim($rowProject["Title"]) . "&pound;",ENT_COMPAT,"UTF-8");
```

 converts the `&pound;` to a £ but it doesn’t touch `&comma;`

How on earth do I convert `&comma;`


Any ideas other than a replace?

Thanks
Pam

**Comment by Knutsford** [comment]
Solved

```
$Questions[] = html_entity_decode(trim($rowQuestion["TheQuestion"]),ENT_COMPAT | ENT_HTML5,"UTF-8");
``` 

etc

The `| ENT_HTML5` Solved it

Hurray
Thank for your help
Pam

**Comment by Knutsford** [comment]
Nope still doesn’t work ☹ I thought you had spotted it- boohoo

```
<html> 

	
<head> 

	
<script type="text/javascript" src="https://gc.kis.v2.scr.kaspersky-labs.com/FD126C42-EBFA-4E12-B309-BB3FDD723AC1/main.js?attr=ivlNFr-4zHMtxCSV0WhozLSlVTO219YeLu1Jcz2rY0NaSM9P5C8WOEk0oQONthBUldDxCJpYwqG-ijIvLN6D7by1Hgun-gx2twdNTJ2sRUPuuXFzpmBstSTFN2uV0zJt" charset="UTF-8"></script><link rel="stylesheet" crossorigin="anonymous" href="https://gc.kis.v2.scr.kaspersky-labs.com/E3E8934C-235A-4B0E-825A-35A08381A191/abn/main.css?attr=aHR0cHM6Ly93d3cuY2RtcmVndWxhdGlvbnMudWsvaW5jbHVkZXMvZ2VuZXJhdGVfbGVnaXNsYXRpb25fY2hlY2tfcGRmLnBocA"/><style> 

	
@import url(http://fonts.googleapis.com/css?family=Dosis); 

	
@import url(http://fonts.googleapis.com/css?family=Carme); 

	
@import url(http://fonts.googleapis.com/css?family=Scada); 

	
@import url(http://fonts.googleapis.com/css?family=Nobile); 

	
@import url(http://fonts.googleapis.com/css?family=Alef); 

	
@import url(http://fonts.googleapis.com/css?family=Economica); 

	
@import url(http://fonts.googleapis.com/css?family=Marmelad); 

	
@import url(http://fonts.googleapis.com/css?family=Viga); 

		
	
@import url(http://fonts.googleapis.com/css?family=Oswald); 

	
@import url(http://fonts.googleapis.com/css?family=Raleway); 

	
@import url(http://fonts.googleapis.com/css?family=Roboto); 

	
@charset "utf-8"; 

		
	
/* ************************** BODY STYLES *****************************/ 

		
	
p 

	
{ 

	
padding-bottom: 12px; 

	
} 

		
	
body { margin: 0px; font-size: 14px; background-color: white;} 

		
	
/* ************************** LIST STYLES *****************************/ 

		
	
ul 

	
{ 

	
list-style-type: square;font-family: "Roboto"; font-size: 11pt; font-weight: 400; color: #0D0B3B; 

	
} 

		
	
/* ************************** HEADER STYLES ****************************/ 

		
	
/* --- HEADER 1 STYLES --- */ 

		
	
@font-face { 

	
font-family: Roboto; 

	
font-weight: 600; 

	
font-style: medium; src: url("https://www.cdmregulations.uk/includes/dompdf/dompdf/lib/fonts/Roboto-Medium.ttf") format("truetype"); 

	
} 

	
@font-face { 

	
font-family: Roboto; 

	
font-weight: 400; 

	
font-style: normal; src: url("https://www.cdmregulations.uk/includes/dompdf/dompdf/lib/fonts/Roboto-Regular.ttf") format("truetype"); 

	
} 

		
	
@font-face { 

	
font-family: Roboto; 

	
font-weight: 900; 

	
font-style: normal; src: url("https://www.cdmregulations.uk/includes/dompdf/dompdf/lib/fonts/Roboto-Black.ttf") format("truetype"); 

	
} 

		
	
strong { 

	
font-family: Roboto; 

	
font-weight: 900; 

	
} 

		
	
P { 

	
font-family: Roboto; 

	
font-weight: 600; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 12pt; 

	
color: #141414; 

	
text-decoration: none; 

	
} 

		
		
	
h1 { 

	
font-family: "Oswald"; 

	
font-weight: 400; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 24pt; 

	
color: #000000; 

	
text-decoration: none; 

	
} 

		
	
h2 { 

	
font-family: "Oswald"; 

	
font-weight: 400; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 18pt; 

	
color: #000000; 

	
text-decoration: none; 

	
} 

	
h2.blue { 

	
font-family: "Oswald"; 

	
font-weight: 400; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 18pt; 

	
color: #00A8E3; 

	
text-decoration: none; 

	
} 

		
	
h3 { 

	
font-family: "Roboto"; 

	
font-weight: 400; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 11pt; 

	
color: #000000; 

	
text-decoration: none; 

	
} 

		
	
h4 { 

	
font-family: "Roboto"; 

	
font-weight: 600; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 12pt; 

	
color: #141414; 

	
text-decoration: none; 

	
} 

		
		
	
h5 { 

	
font-family: "Oswald"; 

	
font-weight: 400; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 16pt; 

	
color: #0092C3; 

	
text-decoration: none; 

	
} 

		
	
h6 { 

	
font-family: "Oswald"; 

	
font-weight: 400; 

	
display: block; 

	
margin: 0px 0 0px 0; 

	
font-size: 18pt; 

	
color: #000000; 

	
text-decoration: none; 

	
} 

	
hr{ 

	
width: 100%; 

	
color: #C0C0C0; 

	
} 

	
style="width: 100%" color="#C0C0C0" size="0" 

		
		
	
/* ************************** TABLE STYLES *****************************/ 

		
	
.NoPads h1, h2, h3, h4, h5, h6 { 

	
padding-bottom: 0; 

	
} 

		
	
div.page_break{ 

	
page-break-before: always; 

	
} 

	
</style> 

	
</head> 

	
<body> 

	
<p align="center"><img src="https://www.cdmregulations.uk/PDF/HEADER_Legislation.jpg" width="100%"></p> 

	
<div align="center"> 

	
<br> 

	
<Br> 

	
<table border="0" style="border-collapse: collapse" width="100%" cellpadding="0"> 

	
<tr> 

	
<td style="text-align: center"> 

	
<h1>Legislation Check</h1> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td style="text-align: center"> 

	
<h2>New boilers</h2> 

	
</td> 

	
</tr> 

	
</table> 

	
</div> 

	
<br><br> 

	
<div align="center"> 

	
<table border="0" width="100%" cellpadding="0" border="1"> 

		
	
<tr> 

	
<td colspan="6"> 

	
<h3>Question 1: Will you be using ladders&comma; steps&comma; podiums or any other equipment that will give you access to height??</h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"> 

	
<h3>Answer: Working at Height Regulations 2005<br /><br />In order to remain compliant&comma; the Working at height regulations require the following to be considered:<br /><br /> 

	
<ul> 

	
<li>Can you avoid working at height?</li> 

	
<li>Are ladders your last resort?</li> 

	
<li>Is the scaffolding being inspected every 7 days?</li> 

	
<li>Are you protecting memebers of public&comma; site operatives from falling objects?&nbsp;</li> 

	
</ul></h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"><h3><br></h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"> 

	
<h3>Question 2: Will you be using power tools such as chop saws&comma; diamond blades and grinders??</h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"> 

	
<h3>Answer: Provision and Use of Work Equipment Regulations (PUWER) 1998</h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"><h3><br></h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"> 

	
<h3>Question 3: Will you be cutting&comma; drilling&comma; chasing&comma; tile cutting??</h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"> 

	
<h3>Answer: Control of Hazardous Substances 2002?</h3> 

	
</td> 

	
</tr> 

	
<tr> 

	
<td colspan="6"><h3><br></h3> 

	
</td> 

	
</tr></table> 

	
</div> 

	
<p align="center"><img src="https://www.cdmregulations.uk/PDF/FOOTER.jpg" width="100%"></p> 

		
	
</body>
```

 

Blast  - Thanks

Pam

**Reply by bsweeney** [reply]
This HTML renders OK for me. There may be an issue in one of the underlying dependencies (e.g., libxml). If you are able I recommend checking for updates to your OS.

Is it possible the entity is being double-encoded somewhere before being passed to Dompdf?

This is not a recommended work around, but you could also try running your HTML through [`html_entity_decode`](https://www.php.net/manual/en/function.html-entity-decode.php) prior to rendering by Dompdf. This would not be my recommended resolution since it may have unintended side effects if you have entities that require encoding (such as `>` or `<`).

**Comment by Knutsford** [comment]
This is what I had

 

 

namespace Dompdf;

require_once 'dompdf/dompdf/autoload.inc.php';

 

 

$dompdf = new Dompdf(); 

 

 

$dompdf->loadHtml('

 

 

etc

 

 

 

I have downloaded the latest version but there is no autoload.inc.php in there

 

What do I do please?

 

Help

Thanks

P Whittaker

**Comment by Knutsford** [comment]
I had downloaded the wrong zip – oops

 

I have now got the latest one downloaded but it still isn’t working

 

 

 

 

 

 

 

 

 

$dompdf->loadHtml('

<html>

    <head>

        <style>

            @import url(http://fonts.googleapis.com/css?family=Dosis);

@import url(http://fonts.googleapis.com/css?family=Carme);

@import url(http://fonts.googleapis.com/css?family=Scada);

@import url(http://fonts.googleapis.com/css?family=Nobile);

@import url(http://fonts.googleapis.com/css?family=Alef);

@import url(http://fonts.googleapis.com/css?family=Economica);

@import url(http://fonts.googleapis.com/css?family=Marmelad);

@import url(http://fonts.googleapis.com/css?family=Viga);

 

@import url(http://fonts.googleapis.com/css?family=Oswald);

@import url(http://fonts.googleapis.com/css?family=Raleway);

@import url(http://fonts.googleapis.com/css?family=Roboto);

@charset "utf-8";

 

/* ************************** BODY STYLES *****************************/

 

p

{

              padding-bottom: 12px;

}

 

body { margin: 0px; font-size: 14px; background-color: white;}

 

/* ************************** LIST STYLES *****************************/

 

ul

{

  list-style-type: square;font-family: "Roboto"; font-size: 11pt; font-weight: 400; color: #0D0B3B;

}

 

/* ************************** HEADER STYLES ****************************/

 

/* --- HEADER 1 STYLES --- */

 

@font-face { 

              font-family: Roboto; 

              font-weight: 600; 

              font-style: medium; src: url("' . $Theprotocol . $_SERVER['HTTP_HOST'] . '/includes/dompdf/dompdf/lib/fonts/Roboto-Medium.ttf") format("truetype"); 

} 

@font-face { 

              font-family: Roboto; 

              font-weight: 400; 

              font-style: normal; src: url("' . $Theprotocol . $_SERVER['HTTP_HOST'] . '/includes/dompdf/dompdf/lib/fonts/Roboto-Regular.ttf") format("truetype"); 

} 

 

@font-face { 

              font-family: Roboto; 

              font-weight: 900;

              font-style: normal; src: url("' . $Theprotocol . $_SERVER['HTTP_HOST'] . '/includes/dompdf/dompdf/lib/fonts/Roboto-Black.ttf") format("truetype"); 

}  

 

strong {

              font-family: Roboto; 

    font-weight: 900;

}

 

P {

              font-family: Roboto;   

               font-weight: 600;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 12pt;

    color: #141414;

    text-decoration: none;

}

 

 

h1 {

    font-family: "Oswald";

    font-weight: 400;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 24pt;

    color: #000000;

    text-decoration: none;

}

 

h2 {

    font-family: "Oswald";

    font-weight: 400;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 18pt;

    color: #000000;

    text-decoration: none;

}

h2.blue {

font-family: "Oswald";

    font-weight: 400;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 18pt;

    color: #00A8E3;

              text-decoration: none;

}

    

h3 {

    font-family: "Roboto";

    font-weight: 400;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 11pt;

    color: #000000;

    text-decoration: none;

}

   

h4 {

    font-family: "Roboto";

    font-weight: 600;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 12pt;

    color: #141414;

    text-decoration: none;

}

 

 

h5 {

    font-family: "Oswald";

    font-weight: 400;

    display: block;

    margin: 0px 0 0px 0;

    font-size: 16pt;

    color: #0092C3;

    text-decoration: none;

}

 

h6 {

    font-family: "Oswald";

    font-weight: 400;

   display: block;

    margin: 0px 0 0px 0;

    font-size: 18pt;

    color: #000000;

    text-decoration: none;

}

hr{

              width: 100%;

              color: #C0C0C0;

}

style="width: 100%" color="#C0C0C0" size="0"

 

 

/* ************************** TABLE STYLES *****************************/

 

.NoPads h1, h2, h3, h4, h5, h6 {

    padding-bottom: 0;

}

 

div.page_break{

    page-break-before: always;

}

        </style>

    </head>

<body

<div align="center">

              <table border="0" width="100%" cellpadding="0" border="1">

                             ' . $Rows

              

               . '</table>

</div>      

</body>

');

 

 

 

If I echo $Rows I get 

 

Question 1: Will you be using power tools such as chop saws, diamond blades and grinders?? 

Answer: Provision and Use of Work Equipment Regulations (PUWER) 1998

 

Question 2: Will you be cutting, drilling, chasing, tile cutting?? 

Answer: Control of Hazardous Substances 2002?

 

 

Which is correct

 

 

 

 

The pdf has

 

Question 1: Will you be using power tools such as chop saws&comma; diamond blades and grinders?? 

Answer: Provision and Use of Work Equipment Regulations (PUWER) 1998 

 

Question 2: Will you be cutting&comma; drilling&comma; chasing&comma; tile cutting?? 

Answer: Control of Hazardous Substances 2002?

 

It still has &comma;

 

 

Thanks

P Whittaker
